src/compiler/checker.ts

```diff
@@ -18458,7 +18458,17 @@ namespace ts {
                     }
                     contextualType = constraint;
                 }
-                return maybeTypeOfKind(contextualType, (TypeFlags.Literal | TypeFlags.Index));

```

Comment:
```ts
return forEachType(contextualType, t => t.flags & (TypeFlags.Literal | TypeFlags.Index));
```

Reply:
@mhegazy That wouldn't accomplish the same thing as my change. `maybeTypeOfKind` actually already did something very similar to what you propose, except that it looked for `UnionOrIntersectionType` for the recursion while `forEachType` only looks for `UnionType`. 

The problem is the behavior from #19632 was caused by not extracting the any generic constraint types during the recursion into the unions and intersections. 

I did go ahead and refactor the code I wrote to use the core forEach function, and also added a few more test cases.